Add Firmata integration docs (#13508)
home-assistant/home-assistant.github.io: source/_integrations/firmata.markdown

---
title: Firmata
description: "Connect Arduino-compatible boards within Home Assistant"
ha_category:
  - DIY
  - Binary Sensor
  - Switch

> ha_release: 0.113

Reviewer: ```suggestion
ha_release: 0.114
```

The Firmata protocol is a standard protocol for microcontrollers. Most of these boards are support digital and analog inputs and outputs. [Arduino](https://www.arduino.cc/) and Arduino-compatible microcontroller development boards are the most popular boards to use with Firmata.

There is currently support for the following device types within Home Assistant:

- [Binary Sensor](#binary_sensor)
- [Switch](#switches)

## Configuration

You will need to upload a [Firmata firmware](https://github.com/firmata/) to your board. Please upload any `StandardFirmata` sketch to your board; please refer to the [Arduino documentation](https://www.arduino.cc/en/Main/Howto) for further information for uploading sketches. Firmata is available in the Arduino Library Manager in the Arduino IDE. You may also use a [Firmata Express](https://github.com/MrYsLab/FirmataExpress).

> To integrate an Arduino boards with Home Assistant, add the following section to your `configuration.yaml` file:

Reviewer 2: ```suggestion
To integrate Arduino boards with Home Assistant, add the following section to your `configuration.yaml` file:
```

<div class='note warning'>

> Firmata does not store last state accross power cycles. This means that with every powerup the pins are set to off/low and after every initialization they will be set to the default that you have configured in YAML.

Reviewer 2: ```suggestion
Firmata does not store the last state across power cycles. This means that with every powerup the pins are set to off/low and after every initialization, they will be set to the default that you have configured in YAML.
```

      required: true
      type: integer
    initial:
      description: The initial output of the pin after initialization. Note that this is inverted if `negate` is enabled.
      required: false
      default: False
      type: boolean
    negate:
      description: Flips the output of the digital pin
      required: false
      default: False
      type: boolean
binary_sensor:
  description: Digital high/low input to configure
  required: false
  type: list
  keys:
    name:
      description: The name of the entity to create in Home Assistant
      required: true
      type: string
    pin:
      description: The digital pin number on the board.
      required: true
      type: integer
    pin_mode:
      description: The digital pin input mode. Supported modes are `INPUT` and `PULLUP`. Check your board specifications to see which pins have optional internal pullups available.
      required: true
      type: integer
    negate:
      description: Flips the input of the digital pin
      required: false
      default: False
      type: boolean
{% endconfiguration %}

<div class='note'>

> NOTE: If you double-configure a pin, the integration will fail to configure the second one that it attempts to setup and will log an error.

Reviewer 2: No need to add a note, as it is already displayed as a note.
```suggestion
If you double-configure a pin, the integration will fail to configure the second one that it attempts to set up and will log an error.
```

---
title: Firmata
description: "Connect Arduino-compatible boards within Home Assistant"
ha_category:
  - DIY
  - Binary Sensor
  - Switch

> ha_release: "0.110"

Reviewer: ```suggestion
ha_release: 0.111
```

ha_iot_class: Local Push

> ha_config_flow: false

Reviewer: If config flow isn't support for now you can remove this line

ha_codeowners:
  - '@DaAwesomeP'
ha_domain: firmata

> redirect_from:

Reviewer: This option is no longer needed, you can remove it.

The Firmata protocol is a standard protocol for microcontrollers. Most of these boards are support digital and analog inputs and outputs. [Arduino](https://www.arduino.cc/) and Arduino-compatible microcontroller development boards are the most popular boards to use with Firmata.

There is currently support for the following device types within Home Assistant:

- [Binary Sensor](#binary_sensor)
- [Switch](#switches)

## Configuration

You will need to upload a [Firmata firmware](https://github.com/firmata/) to your board. Please upload any `StandardFirmata` sketch to your board; please refer to the [Arduino documentation](https://www.arduino.cc/en/Main/Howto) for further information for uploading sketches. Firmata is available in the Arduino Library Manager in the Arduino IDE. You may also use a [Firmata Express](https://github.com/MrYsLab/FirmataExpress).

To integrate an Arduino boards with Home Assistant, add the following section to your `configuration.yaml` file:

> 

Reviewer: ```suggestion
```

Please remove the extra blank lines (except 1 :smile: )

<div class='note warning'>
Firmata does not store last state accross power cycles. This means that with every powerup the pins are set to off/low and after every initialization they will be set to the default that you have configured in YAML.
</div>

<div class='note warning'>

> Firmata may cycle pins HIGH/LOW during initialization. If your pins should not be accidentally cycled, you may need to modify your Firmtata sketch or if you are using a relay board use an output pin as VCC. See this issue for details: <a href="https://github.com/firmata/arduino/issues/166" target="_blank">Make initial pin value configurable</a>

Reviewer: It's also possible to use the markdown style for the link to Github: [See our docs](https://developers.home-assistant.io/docs/documenting/create-page#html)

Author: I tried but it doesn't seem to work inside the `<div>` unless I typoed it. I will try again.

Author: Oh I see! I need a newline!
